### PR TITLE
BACK-1006 Transactor : create Transaction for signature

### PR DESCRIPTION
**Pull request by estrauser-ledger**
## What is this about?

This is the first part of transaction creation, with UTXO picking

## Cute picture of lama
footlama

### CODE REVIEW

**Comment by estrauser-ledger** [comment]
One it:test doesn't work since I used the new InterpreterServiceMock in the CursorStateServiceIT.
I'll have to check what's going on..

Also stuff that's needed to make it work :
- create tx v2 after signing
- dust management
- change creation
- fee estimation
- broadcast

File: coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/routes/AccountController.scala [hunk 1]
```diff
@@ -213,6 +216,18 @@ object AccountController extends Http4sDsl[IO] with IOLogging {
...
+          ) / "transaction" =>
+        log.info(s"preparing transaction creation for account: $accountId") *>
+          transactorClient
+            .prepareTransaction(
+              accountId,
+              CoinSelector.DepthFirst,
+              List()
+            )
+            .flatMap(Ok(_))
     }
```
Comment (estrauser-ledger) [inline coins/bitcoin/api/src/main/scala/co/ledger/lama/bitcoin/api/routes/AccountController.scala:231]: Not done properly, missing body parsing

File: coins/bitcoin/common/src/main/protobuf/transactor.proto [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+
+message PrepareTransactionRequest {
+  bytes accountId = 1;
+  CoinSelector pickingStrategy = 2;
+  repeated PrepareTxOutput outputs = 3;
+}
+
+message PrepareTransactionResponse {
+  string hex = 1;
+}
```
Comment (estrauser-ledger) [inline coins/bitcoin/common/src/main/protobuf/transactor.proto:25]: Add utxos, script and withnessHash (last 2 are returned by btcd)

File: coins/bitcoin/common/src/main/scala/co/ledger/lama/bitcoin/common/models/transactor/package.scala [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+import io.circe.{Decoder, Encoder}
+import co.ledger.lama.bitcoin.transactor.protobuf
+
+object transactor {
+
+  abstract class CoinSelector(val name: String) {
+    def toProto: protobuf.CoinSelector
+  }
+
+  object CoinSelector {
```
Comment (estrauser-ledger) [inline coins/bitcoin/common/src/main/scala/co/ledger/lama/bitcoin/common/models/transactor/package.scala:12]: Maybe rename PickingStrategy ? What do you think @nukky94

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+    //TODO: log
+    for {
+      accountId <- UuidUtils.bytesToUuidIO(request.accountId)
+      utxos     <- getUTXOs(accountId, 100, Sort.Ascending).compile.toList
+      outputs = request.outputs.map(PrepareTxOutput.fromProto(_)).toList
+      pickedUtxos <- PickingStrategyService.pickUtxos(
+        CoinSelector.fromProto(request.pickingStrategy),
+        utxos,
+        outputs.map(_.value).sum
+      )
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala:43]: could do this with Stream until the picking

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+      BitcoinNetwork.MainNet
+    )
+  }
+
+  private def utxosToInputs(utxo: Utxo): bitcoinLib.Input = {
+    bitcoinLib.Input(
+      utxo.transactionHash,
+      utxo.outputIndex,
+      utxo.scriptHex,
+      Int.MaxValue * 2 //sequence is actually a unsigned Int in btcd
```
Comment (estrauser-ledger) [inline coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala:100]: @nukky94 didn't you tell me the sequence was to be handled by the bitcoinLib ?

File: coins/bitcoin/worker/src/it/scala/co/ledger/lama/bitcoin/worker/KeychainClientServiceMock.scala [hunk 6]
```diff
@@ -1,10 +1,9 @@
-package co.ledger.lama.bitcoin.worker
+package co.ledger.lama.bitcoin.common.services
```
Comment (estrauser-ledger) [inline coins/bitcoin/worker/src/it/scala/co/ledger/lama/bitcoin/worker/KeychainClientServiceMock.scala:1]: ?

**Review by nukky94 — APPROVED** [review]

File: coins/bitcoin/api/src/main/resources/application.conf [hunk 7]
```diff
@@ -76,14 +76,14 @@ bitcoin {
...
-    host = ${?BITCOIN_BROADCASTER_HOST}
+    host = ${?BITCOIN_TRANSACTOR_HOST}
 
     port = 9997
-    port = ${?BITCOIN_BROADCASTER_PORT}
+    port = ${?BITCOIN_TRANSACTOR_PORT}
 
     ssl = false
-    ssl = ${?BITCOIN_BROADCASTER_SSL}
+    ssl = ${?BITCOIN_TRANSACTOR_SSL}
```
[the remark this answers is not in the document]
Comment (estrauser-ledger) [reply coins/bitcoin/api/src/main/resources/application.conf:87]: :+1:

File: coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala (hunk 4 above)
Comment (estrauser-ledger) [reply coins/bitcoin/transactor/src/main/scala/co/ledger/lama/bitcoin/transactor/Transactor.scala]: https://ledgerhq.atlassian.net/browse/BACK-1044

**Pull request merged**